File: frontends/Python/exasim/Preprocessing/uniformrefinement.py
import numpy as np

from .createdgnodes import createdgnodes
from .facenumbering import facenumbering
from .mkshape import mkshape


def uniformrefinemesh(mesh, app, master, elem2cpu=None):
    """Uniformly refine a Python frontend mesh in parent-major order."""
    nlevel = int(app.get('uniformrefinementlevel', 0))
    if nlevel == 0:
        return mesh, elem2cpu
    if nlevel < 0 or nlevel != app.get('uniformrefinementlevel', 0):
        raise ValueError("uniformrefinementlevel must be a nonnegative integer.")
    if 'uhat' in mesh and len(mesh['uhat']) > 0:
        raise ValueError("uniformrefinementlevel > 0 cannot prolongate the face-based mesh['uhat'] field.")

    if _isempty(mesh.get('dgnodes', [])):
        try:
            f = facenumbering(mesh['p'], mesh['t'], app['elemtype'], mesh['boundaryexpr'], mesh['periodicexpr'])[0]
            mesh['dgnodes'] = createdgnodes(mesh['p'], mesh['t'], f, mesh['curvedboundary'],
                                            mesh['curvedboundaryexpr'], app['porder'])
        except Exception:
            pass

    T = _refinement_template(app['nd'], app['elemtype'], app['nve'])
    R = _refinement_prolongation(T, master['xpe'], app['porder'])

    ne0 = mesh['t'].shape[1]
    np0 = mesh['p'].shape[1]
    for name in ['dgnodes', 'udg', 'vdg', 'wdg']:
        _check_field(mesh, name, master['xpe'].shape[0], ne0)
    if elem2cpu is not None:
        elem2cpu = np.asarray(elem2cpu, dtype=int).reshape(-1, order='F')
        if elem2cpu.size == 0:
            elem2cpu = None
        elif elem2cpu.size != ne0:
            raise ValueError("uniformrefinement: elem2cpu must contain one entry per parent element.")

    for _ in range(nlevel):
        mesh['p'], mesh['t'] = _refine_connectivity(mesh['p'], mesh['t'], T, R, mesh.get('dgnodes', []))
        for name in ['dgnodes', 'udg', 'vdg', 'wdg']:
            _prolong_mesh_field(mesh, name, R['P'])
        if elem2cpu is not None:
            elem2cpu = np.repeat(elem2cpu, T['nchild'])

    print("uniformrefinementlevel = %d: ne %d -> %d, np %d -> %d" %
          (nlevel, ne0, mesh['t'].shape[1], np0, mesh['p'].shape[1]))
    return mesh, elem2cpu


def coarseelem2cpu(dmd, ne):
    elem2cpu = -np.ones(ne, dtype=int)
    for i, part in enumerate(dmd):
        owned = int(np.sum(np.asarray(part['elempartpts']).reshape(-1, order='F')[:2]))
        elem = np.asarray(part['elempart']).reshape(-1, order='F')[:owned]
        elem2cpu[elem] = i
    if np.any(elem2cpu < 0):
        raise ValueError("coarseelem2cpu: some coarse elements do not have an owner rank.")
    return elem2cpu


def _refinement_template(nd, elemtype, nve):
    xlat = []
    children = []

    def addpt(*x):
        xlat.append(tuple(x[:nd]))

    def addchild(v):
        children.append(tuple(v))

    if nd == 1 or elemtype == 1:
        ny = 3 if nd >= 2 else 1
        nz = 3 if nd == 3 else 1
        for k in range(nz):
            for j in range(ny):
                for i in range(3):
                    addpt(0.5*i, 0.5*j, 0.5*k)
        lat = lambda i, j, k: i + 3*(j + 3*k)
        if nd == 1:
            addchild([lat(0, 0, 0), lat(1, 0, 0)])
            addchild([lat(1, 0, 0), lat(2, 0, 0)])
        elif nd == 2:
            for cy in range(2):
                for cx in range(2):
                    addchild([lat(cx, cy, 0), lat(cx+1, cy, 0),
                              lat(cx+1, cy+1, 0), lat(cx, cy+1, 0)])
        else:
            for cz in range(2):
                for cy in range(2):
                    for cx in range(2):
                        addchild([lat(cx, cy, cz), lat(cx+1, cy, cz),
                                  lat(cx+1, cy+1, cz), lat(cx, cy+1, cz),
                                  lat(cx, cy, cz+1), lat(cx+1, cy, cz+1),
                                  lat(cx+1, cy+1, cz+1), lat(cx, cy+1, cz+1)])
    elif nd == 2:
        addpt(0, 0, 0); addpt(1, 0, 0); addpt(0, 1, 0)
        addpt(0.5, 0, 0); addpt(0.5, 0.5, 0); addpt(0, 0.5, 0)
        addchild([0, 3, 5])
        addchild([3, 1, 4])
        addchild([5, 4, 2])
        addchild([3, 4, 5])
    elif nd == 3:
        addpt(0, 0, 0); addpt(1, 0, 0); addpt(0, 1, 0); addpt(0, 0, 1)
        addpt(0.5, 0, 0); addpt(0, 0.5, 0); addpt(0, 0, 0.5)
        addpt(0.5, 0.5, 0); addpt(0.5, 0, 0.5); addpt(0, 0.5, 0.5)
        addchild([0, 4, 5, 6])
        addchild([4, 1, 7, 8])
        addchild([5, 7, 2, 9])
        addchild([6, 8, 9, 3])
        addchild([4, 5, 6, 8])
        addchild([4, 5, 7, 8])
        addchild([5, 6, 8, 9])
        addchild([5, 7, 8, 9])

    child = np.asarray(children, dtype=int).T
    if child.size != nve*child.shape[1]:
        raise ValueError("uniformrefinement: unsupported element.")
    xlat = np.asarray(xlat, dtype=float)
    phi = _linear_basis(xlat, nd, elemtype)
    nsup = np.zeros(xlat.shape[0], dtype=int)
    sup = np.zeros((nve, xlat.shape[0]), dtype=int)
    for l in range(xlat.shape[0]):
        s = np.where(phi[l, :] > 1e-12)[0]
        nsup[l] = s.size
        sup[:s.size, l] = s
    if elemtype == 0 and nd >= 2:
        for c in range(child.shape[1]):
            cv = child[:, c]
            E = xlat[cv[1:nd+1], :] - xlat[cv[0], :]
            if np.linalg.det(E[:, :nd]) < 0:
                child[[nd-1, nd], c] = child[[nd, nd-1], c]
    return {'nd': nd, 'elemtype': elemtype, 'nve': nve, 'xlat': xlat,
            'nlat': xlat.shape[0], 'nchild': child.shape[1],
            'child': child, 'nsup': nsup, 'sup': sup}


def _refinement_prolongation(T, xpe, porder):
    npe = xpe.shape[0]
    phielem = _linear_basis(xpe, T['nd'], T['elemtype'])
    P = np.zeros((npe, npe, T['nchild']))
    Slat = mkshape(porder, xpe, T['xlat'], T['elemtype'])[:, :, 0]
    for c in range(T['nchild']):
        pts = np.zeros((npe, T['nd']))
        for d in range(T['nd']):
            for v in range(T['nve']):
                pts[:, d] += phielem[:, v]*T['xlat'][T['child'][v, c], d]
        shap = mkshape(porder, xpe, pts, T['elemtype'])[:, :, 0]
        P[:, :, c] = shap.T
        if np.max(np.abs(np.sum(P[:, :, c], axis=1) - 1.0)) > 1e-8:
            raise ValueError("uniformrefinement: prolongation is not a partition of unity.")
    return {'P': P, 'Slat': Slat}


def _refine_connectivity(p, t, T, R, dgnodes):
    nd, npv = p.shape
    ne = t.shape[1]
    pref = np.array(p, copy=True, order='F')
    tref = np.zeros((T['nve'], ne*T['nchild']), dtype=t.dtype, order='F')
    index = {}
    latid = np.zeros(T['nlat'], dtype=int)
    has_dgnodes = not _isempty(dgnodes)
    for e in range(ne):
        te = t[:, e]
        for l in range(T['nlat']):
            ns = T['nsup'][l]
            if ns == 1:
                latid[l] = te[T['sup'][0, l]]
                continue
            support = tuple(sorted(int(te[v]) for v in T['sup'][:ns, l]))
            if support not in index:
                new_id = npv
                npv += 1
                index[support] = new_id
                if has_dgnodes:
                    x = np.array([np.dot(R['Slat'][:, l], dgnodes[:, d, e]) for d in range(nd)])
                else:
                    x = np.mean(p[:, list(support)], axis=1)
                pref = np.column_stack((pref, x))
            latid[l] = index[support]
        for c in range(T['nchild']):
            tref[:, e*T['nchild'] + c] = latid[T['child'][:, c]]
    return np.asarray(pref, order='F'), tref


def _prolong_mesh_field(mesh, name, P):
    if name not in mesh or _isempty(mesh[name]):
        return
    f = mesh[name]
    npe, nc, ne = f.shape
    nchild = P.shape[2]
    g = np.zeros((npe, nc, ne*nchild), dtype=f.dtype, order='F')
    for e in range(ne):
        for c in range(nchild):
            g[:, :, e*nchild + c] = P[:, :, c] @ f[:, :, e]
    mesh[name] = g


def _check_field(mesh, name, npe, ne):
    if name not in mesh or _isempty(mesh[name]):
        return
    if mesh[name].shape[0] != npe or mesh[name].shape[2] != ne:
        raise ValueError("uniformrefinement: mesh['%s'] must have size npe x ncomp x ne." % name)


def _isempty(a):
    return a is None or (isinstance(a, list) and len(a) == 0) or np.size(a) == 0


def _linear_basis(pts, nd, elemtype):
    if nd == 1:
        xi = pts[:, 0]
        return np.column_stack((1-xi, xi))
    if nd == 2 and elemtype == 0:
        xi = pts[:, 0]; eta = pts[:, 1]
        return np.column_stack((1-xi-eta, xi, eta))
    if nd == 2 and elemtype == 1:
        xi = pts[:, 0]; eta = pts[:, 1]
        return np.column_stack(((1-xi)*(1-eta), xi*(1-eta), xi*eta, (1-xi)*eta))
    if nd == 3 and elemtype == 0:
        xi = pts[:, 0]; eta = pts[:, 1]; zeta = pts[:, 2]
        return np.column_stack((1-xi-eta-zeta, xi, eta, zeta))
    if nd == 3 and elemtype == 1:
        xi = pts[:, 0]; eta = pts[:, 1]; zeta = pts[:, 2]
        return np.column_stack(((1-xi)*(1-eta)*(1-zeta), xi*(1-eta)*(1-zeta),
                                xi*eta*(1-zeta), (1-xi)*eta*(1-zeta),
                                (1-xi)*(1-eta)*zeta, xi*(1-eta)*zeta,
                                xi*eta*zeta, (1-xi)*eta*zeta))
    raise ValueError("uniformrefinement: unsupported element.")

File: frontends/Julia/Exasim/src/Preprocessing/uniformrefinement.jl
function uniformrefinemesh(mesh, app, master, elem2cpu=Int[])

nlevel = app.uniformrefinementlevel;
if nlevel == 0
    return mesh, elem2cpu;
end
if nlevel < 0
    error("uniformrefinementlevel must be a nonnegative integer.");
end

if size(mesh.dgnodes, 3) == 0
    try
        f,~,~ = facenumbering(mesh.p, mesh.t, app.elemtype, mesh.boundaryexpr, mesh.periodicexpr);
        mesh.dgnodes = createdgnodes(mesh.p, mesh.t, f, mesh.curvedboundary, mesh.curvedboundaryexpr, app.porder);
    catch
    end
end

T = refinementtemplate(app.nd, app.elemtype, size(mesh.t, 1));
R = refinementprolongation(T, master.xpe, app.porder);

ne0 = size(mesh.t, 2);
np0 = size(mesh.p, 2);
checkfield(mesh.dgnodes, "dgnodes", size(master.xpe, 1), ne0);
checkfield(mesh.udg, "udg", size(master.xpe, 1), ne0);
checkfield(mesh.odg, "odg", size(master.xpe, 1), ne0);
checkfield(mesh.wdg, "wdg", size(master.xpe, 1), ne0);
if !isempty(elem2cpu) && length(elem2cpu) != ne0
    error("uniformrefinement: elem2cpu must contain one entry per parent element.");
end

for level = 1:nlevel
    xdg = size(mesh.dgnodes, 3) > 0 ? mesh.dgnodes : zeros(Float64, 0, 0, 0);
    mesh.p, mesh.t = refineconnectivity(mesh.p, mesh.t, T, R, xdg);
    mesh.dgnodes = prolongfield(mesh.dgnodes, R.P);
    mesh.udg = prolongfield(mesh.udg, R.P);
    mesh.odg = prolongfield(mesh.odg, R.P);
    mesh.wdg = prolongfield(mesh.wdg, R.P);
    if !isempty(elem2cpu)
        elem2cpu = reshape(repeat(reshape(elem2cpu, 1, :), T.nchild, 1), :);
    end
end

println("uniformrefinementlevel = $nlevel: ne $ne0 -> $(size(mesh.t, 2)), np $np0 -> $(size(mesh.p, 2))");
return mesh, elem2cpu;

end

function coarseelem2cpu(dmd, ne)

elem2cpu = -ones(Int, ne);
for i = 1:length(dmd)
    owned = sum(dmd[i].elempartpts[1:min(2, length(dmd[i].elempartpts))]);
    elem = dmd[i].elempart[1:owned];
    elem2cpu[elem] .= i - 1;
end
if any(elem2cpu .< 0)
    error("coarseelem2cpu: some coarse elements do not have an owner rank.");
end
return elem2cpu;

end

function refinementtemplate(nd, elemtype, nve)

xlat = Array{Float64,2}(undef, 0, nd);
children = Vector{Vector{Int}}();
addpt(x...) = (xlat = [xlat; reshape(Float64[x[i] for i=1:nd], 1, nd)]);
function addchild(v)
    push!(children, Int.(v));
end

if nd == 1 || elemtype == 1
    ny = nd >= 2 ? 3 : 1;
    nz = nd == 3 ? 3 : 1;
    for k = 0:(nz-1), j = 0:(ny-1), i = 0:2
        addpt(0.5*i, 0.5*j, 0.5*k);
    end
    lat(i,j,k) = 1 + i + 3*(j + 3*k);
    if nd == 1
        addchild([lat(0,0,0), lat(1,0,0)]);
        addchild([lat(1,0,0), lat(2,0,0)]);
    elseif nd == 2
        for cy = 0:1, cx = 0:1
            addchild([lat(cx,cy,0), lat(cx+1,cy,0), lat(cx+1,cy+1,0), lat(cx,cy+1,0)]);
        end
    else
        for cz = 0:1, cy = 0:1, cx = 0:1
            addchild([lat(cx,cy,cz), lat(cx+1,cy,cz), lat(cx+1,cy+1,cz), lat(cx,cy+1,cz),
                      lat(cx,cy,cz+1), lat(cx+1,cy,cz+1), lat(cx+1,cy+1,cz+1), lat(cx,cy+1,cz+1)]);
        end
    end
elseif nd == 2
    addpt(0, 0, 0); addpt(1, 0, 0); addpt(0, 1, 0);
    addpt(0.5, 0, 0); addpt(0.5, 0.5, 0); addpt(0, 0.5, 0);
    addchild([1, 4, 6]);
    addchild([4, 2, 5]);
    addchild([6, 5, 3]);
    addchild([4, 5, 6]);
elseif nd == 3
    addpt(0, 0, 0); addpt(1, 0, 0); addpt(0, 1, 0); addpt(0, 0, 1);
    addpt(0.5, 0, 0); addpt(0, 0.5, 0); addpt(0, 0, 0.5);
    addpt(0.5, 0.5, 0); addpt(0.5, 0, 0.5); addpt(0, 0.5, 0.5);
    addchild([1, 5, 6, 7]);
    addchild([5, 2, 8, 9]);
    addchild([6, 8, 3, 10]);
    addchild([7, 9, 10, 4]);
    addchild([5, 6, 7, 9]);
    addchild([5, 6, 8, 9]);
    addchild([6, 7, 9, 10]);
    addchild([6, 8, 9, 10]);
end

nchild = length(children);
if nchild == 0
    error("uniformrefinement: unsupported element.");
end
child = hcat(children...);
if length(child) != nve*nchild
    error("uniformrefinement: unsupported element.");
end

phi = linearbasis(xlat, nd, elemtype);
nlat = size(xlat, 1);
nsup = zeros(Int, nlat);
sup = zeros(Int, nve, nlat);
for l = 1:nlat
    s = findall(phi[l, :] .> 1e-12);
    nsup[l] = length(s);
    sup[1:length(s), l] = s;
end
if elemtype == 0 && nd >= 2
    for c = 1:nchild
        cv = child[:, c];
        E = xlat[cv[2:nd+1], :] .- xlat[cv[1], :]';
        if localdet(E[:, 1:nd]) < 0
            child[[nd, nd+1], c] = child[[nd+1, nd], c];
        end
    end
end
return (nd=nd, elemtype=elemtype, nve=nve, xlat=xlat, nlat=nlat,
        nchild=nchild, child=child, nsup=nsup, sup=sup);

end

function localdet(A)

if size(A, 1) == 2
    return A[1,1]*A[2,2] - A[1,2]*A[2,1];
elseif size(A, 1) == 3
    return A[1,1]*(A[2,2]*A[3,3] - A[2,3]*A[3,2]) -
           A[1,2]*(A[2,1]*A[3,3] - A[2,3]*A[3,1]) +
           A[1,3]*(A[2,1]*A[3,2] - A[2,2]*A[3,1]);
end
error("uniformrefinement: unsupported determinant size.");

end

function refinementprolongation(T, xpe, porder)

npe = size(xpe, 1);
phielem = linearbasis(xpe, T.nd, T.elemtype);
P = zeros(Float64, npe, npe, T.nchild);
Slat = mkshape(porder, xpe, T.xlat, T.elemtype)[:, :, 1];
for c = 1:T.nchild
    pts = zeros(Float64, npe, T.nd);
    for d = 1:T.nd, v = 1:T.nve
        pts[:, d] += phielem[:, v]*T.xlat[T.child[v, c], d];
    end
    shap = mkshape(porder, xpe, pts, T.elemtype)[:, :, 1];
    P[:, :, c] = shap';
    if maximum(abs.(sum(P[:, :, c], dims=2) .- 1.0)) > 1e-8
        error("uniformrefinement: prolongation is not a partition of unity.");
    end
end
return (P=P, Slat=Slat);

end

function refineconnectivity(p, t, T, R, dgnodes)

nd,np = size(p);
ne = size(t, 2);
pref = copy(p);
tref = zeros(Int, T.nve, ne*T.nchild);
index = Dict{String,Int}();
latid = zeros(Int, T.nlat);
hasdgnodes = size(dgnodes, 3) > 0;

for e = 1:ne
    te = t[:, e];
    for l = 1:T.nlat
        ns = T.nsup[l];
        if ns == 1
            latid[l] = te[T.sup[1, l]];
            continue;
        end
        support = sort(te[T.sup[1:ns, l]]);
        key = join(support, "_");
        if haskey(index, key)
            latid[l] = index[key];
        else
            id = np + 1;
            np = id;
            index[key] = id;
            if hasdgnodes
                x = [sum(R.Slat[:, l].*dgnodes[:, d, e]) for d=1:nd];
            else
                x = vec(mean(p[:, support], dims=2));
            end
            pref = [pref x];
            latid[l] = id;
        end
    end
    for c = 1:T.nchild
        tref[:, (e-1)*T.nchild + c] = latid[T.child[:, c]];
    end
end
return pref, tref;

end

function prolongfield(f, P)

if size(f, 3) == 0
    return f;
end
npe,nc,ne = size(f);
nchild = size(P, 3);
g = zeros(eltype(f), npe, nc, ne*nchild);
for e = 1:ne, c = 1:nchild
    g[:, :, (e-1)*nchild + c] = P[:, :, c]*f[:, :, e];
end
return g;

end

function checkfield(f, name, npe, ne)

if size(f, 3) == 0
    return;
end
if size(f, 1) != npe || size(f, 3) != ne
    error("uniformrefinement: mesh.$name must have size npe x ncomp x ne.");
end

end

function linearbasis(pts, nd, elemtype)

if nd == 1
    xi = pts[:, 1];
    return [1.0 .- xi xi];
elseif nd == 2 && elemtype == 0
    xi = pts[:, 1]; eta = pts[:, 2];
    return [1.0 .- xi .- eta xi eta];
elseif nd == 2 && elemtype == 1
    xi = pts[:, 1]; eta = pts[:, 2];
    return [(1.0 .- xi).*(1.0 .- eta) xi.*(1.0 .- eta) xi.*eta (1.0 .- xi).*eta];
elseif nd == 3 && elemtype == 0
    xi = pts[:, 1]; eta = pts[:, 2]; zeta = pts[:, 3];
    return [1.0 .- xi .- eta .- zeta xi eta zeta];
elseif nd == 3 && elemtype == 1
    xi = pts[:, 1]; eta = pts[:, 2]; zeta = pts[:, 3];
    return [(1.0 .- xi).*(1.0 .- eta).*(1.0 .- zeta) xi.*(1.0 .- eta).*(1.0 .- zeta) xi.*eta.*(1.0 .- zeta) (1.0 .- xi).*eta.*(1.0 .- zeta) (1.0 .- xi).*(1.0 .- eta).*zeta xi.*(1.0 .- eta).*zeta xi.*eta.*zeta (1.0 .- xi).*eta.*zeta];
end
error("uniformrefinement: unsupported element.");

end
